nutils/evaluable.py

```diff
@@ -757,10 +757,30 @@ def _assparse(self):
       return chunks
     return _assparse
 
+  class _evalf_checker:
+    def __init__(self, orig):
+      self.evalf_obj = getattr(orig, '__get__', lambda *args: orig)
+    def __get__(self, instance, owner):
+      evalf = self.evalf_obj(instance, owner)
+      @functools.wraps(evalf)
+      def evalf_with_check(*args, **kwargs):
+        res = evalf(*args, **kwargs)
+        if hasattr(instance, 'dtype'):
+          assert numpy.can_cast(res.dtype, instance.dtype, casting='safe'), ((instance.dtype, res.dtype), instance, res)
```

Comment:
I don't really see why we would do a can cast check instead of equality check. I'm exaggerating here, but what withholds us from setting all dtypes to complex? Then there is the following problem. The behavior of `numpy.ndarray.__getitem__` depends on the dtype of the argument:

```python
numpy.array([1,2,3])[numpy.array([0,1,0],int)] # array([1, 2, 1])
numpy.array([1,2,3])[numpy.array([0,1,0],bool)] # array([2])
```

I didn't check, but I don't think this is something we're currently doing in any of our `Array.evalf` s. There could be other functions that have similar behavior. Who knows; it's Numpy. However, if we enforce strict equality, we don't even have to worry about this.

Reply:
When I suggested before that 'same_kind' is our mode I overlooked that the description includes safe casts, sorry about that. There should be no confusion that the return value should adhere to the Array's dtype.